### PR TITLE
[GlobalIsel] combine ext of trunc with flags

### PR DESCRIPTION
**Pull request by tschuett**
https://github.com/llvm/llvm-project/pull/85592

https://discourse.llvm.org/t/rfc-add-nowrap-flags-to-trunc/77453

https://github.com/llvm/llvm-project/pull/88609

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -7138,3 +7138,69 @@ bool CombinerHelper::matchAddOverflow(MachineInstr &MI, BuildFnTy &MatchInfo) {
 
   return false;
 }
+
+void CombinerHelper::applyBuildFnMO(const MachineOperand &MO,
+                                    BuildFnTy &MatchInfo) {
+  MachineInstr *Root = getDefIgnoringCopies(MO.getReg(), MRI);
+  Builder.setInstrAndDebugLoc(*Root);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7145]: This should be unnecessary since 360f7f5674c4ddef006d233ef7259f8fe07ae81b

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 2]
```diff
@@ -7138,3 +7138,69 @@ bool CombinerHelper::matchAddOverflow(MachineInstr &MI, BuildFnTy &MatchInfo) {
...
+    return false;
+
+  GTrunc *Trunc = getOpcodeDef<GTrunc>(Sext->getSrcReg(), MRI);
+  if (!Trunc)
+    return false;
+
+  // The trunc must have the nsw flag.
+  if (!Trunc->getFlag(MachineInstr::MIFlag::NoSWrap))
+    return false;
+
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7163]: All of this is already in the pattern?  Don't need to double check it?

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-with-flags.mir [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    ; CHECK-LABEL: name: sext_trunc_nsw_nuw
+    ; CHECK: liveins: $w0, $w1
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:_(s64) = COPY $x0
+    ; CHECK-NEXT: $x1 = COPY [[COPY]](s64)
+    %0:_(s64) = COPY $x0
+    %2:_(s32) = nsw nuw G_TRUNC %0
+    %3:_(s64) = G_SEXT  %2
+    $x1 = COPY %3
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-with-flags.mir:156]: Test vector cases

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-with-flags.mir [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+name:            zext_trunc_nuw_vector
+body:             |
+  bb.0:
+    liveins: $w0, $w1
+    ; CHECK-LABEL: name: zext_trunc_nuw_vector
+    ; CHECK: liveins: $w0, $w1
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:_(s32) = COPY $w0
+    ; CHECK-NEXT: [[COPY1:%[0-9]+]]:_(s32) = COPY $w1
+    ; CHECK-NEXT: %bv0:_(<4 x s32>) = G_BUILD_VECTOR [[COPY]](s32), [[COPY1]](s32), [[COPY]](s32), [[COPY1]](s32)
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-with-flags.mir:63]: Vector test.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 5]
```diff
@@ -7138,3 +7138,69 @@ bool CombinerHelper::matchAddOverflow(MachineInstr &MI, BuildFnTy &MatchInfo) {
...
+  Register Dst = Sext->getReg(0);
+  Register Src = Trunc->getSrcReg();
+
+  LLT DstTy = MRI.getType(Dst);
+  LLT SrcTy = MRI.getType(Src);
+
+  // The types have to match for a no-op.
+  if (DstTy != SrcTy)
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7171]: This is the important check. We need a way to determine the input and destination type.

**Comment by tschuett** [comment]
I can remove the flags tests. We need a mechanism to determine the type of `x` in `ext(trunc(x))`. The Sext, Zext, and Trunk wrappers make my live easier to walk over the expression.

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-with-flags.mir [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    liveins: $w0, $w1
+    ; CHECK-LABEL: name: sext_trunc_nsw_types_wrong
+    ; CHECK: liveins: $w0, $w1
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:_(s64) = COPY $x0
+    ; CHECK-NEXT: [[TRUNC:%[0-9]+]]:_(s16) = nsw G_TRUNC [[COPY]](s64)
+    ; CHECK-NEXT: [[SEXT:%[0-9]+]]:_(s32) = G_SEXT [[TRUNC]](s16)
+    ; CHECK-NEXT: $w1 = COPY [[SEXT]](s32)
+    %0:_(s64) = COPY $x0
+    %2:_(s16) = nsw G_TRUNC %0
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-with-flags.mir:138]: This test failed because the types s64 and s32 don't match. The correct flag nsw was set.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 7]
```diff
@@ -7146,3 +7138,60 @@ bool CombinerHelper::matchAddOverflow(MachineInstr &MI, BuildFnTy &MatchInfo) {
...
+
+bool CombinerHelper::matchSextOfTrunc(const MachineOperand &MO,
+                                      BuildFnTy &MatchInfo) {
+  GSext *Sext = getOpcodeDef<GSext>(MO.getReg(), MRI);
+  if (!Sext)
+    return false;
+
+  GTrunc *Trunc = getOpcodeDef<GTrunc>(Sext->getSrcReg(), MRI);
+  if (!Trunc)
+    return false;
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7157]: Should be able to just assert these?

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp]: The asserts are know in the `cast`. It is like I *know* that it will be a zext.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp (hunk 5 above)
Comment (arsenm) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7161]: This also probably needs a legality check

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7161]: We just do a copy, when we reach the copy the types of the two registers are equal.

**Comment by tschuett** [comment]
https://github.com/llvm/llvm-project/blob/4dd103e9c65de7d3dbf12e76fbb72724127ec325/llvm/include/llvm/Target/GlobalISel/Combine.td#L718
should become
```
def zext_trunc_kb : GICombineRule<
   (defs root:$root, build_fn_matchinfo:$matchinfo),
   (match (G_TRUNC $src, $x, (MIFlags not NoUWrap)),
          (G_ZEXT $root, $src),
```

File: llvm/lib/CodeGen/GlobalISel/GISelKnownBits.cpp [hunk 8]
```diff
@@ -64,8 +64,11 @@ KnownBits GISelKnownBits::getKnownBits(MachineInstr &MI) {
 
 KnownBits GISelKnownBits::getKnownBits(Register R) {
   const LLT Ty = MRI.getType(R);
+  // Since the number of lanes in a scalable vector is unknown at compile time,
+  // we track one bit which is implicitly broadcast to all lanes.  This means
+  // that all lanes in a scalable vector are considered demanded.
   APInt DemandedElts =
-      Ty.isVector() ? APInt::getAllOnes(Ty.getNumElements()) : APInt(1, 1);
+      Ty.isFixedVector() ? APInt::getAllOnes(Ty.getNumElements()) : APInt(1, 1);
```
Comment (Pierre-vh) [inline llvm/lib/CodeGen/GlobalISel/GISelKnownBits.cpp:71]: Should this be a separate patch? What's the impact?

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/GISelKnownBits.cpp:71]: There is a G_SPLAT_VECTOR test and a second combine zext_trunc_fold_matchinfo that relies on known bits to prove that zext(trunc(x)) is a noop. It crashed.

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/GISelKnownBits.cpp:71]: The changes a copy-pasted from the DAG to support scalable vectors in known bits.

**Comment by tschuett** [comment]
Ping.

**Review by arsenm — APPROVED** [review]
